tests/unit/test_cmr_search.py

```diff
@@ -238,4 +249,87 @@ def test_exceptions_granule_link(self):
         with self.subTest('Granule has `links` and all correct attributes'):
             self.assertEqual(get_granule_link(granule_response_links_correct),
                              'https://data.gesdisc.earthdata.nasa.gov/.nc4')
-    
\ No newline at end of file
+    
+    def test_granule_name(self):
+        ''' Check if get_granule_name returns expected string for response
+            with key "producer_granule_id" and raises IndexError for response
+            without a "producer_granule_id"
+        '''
+        granule_response = [{
+            'producer_granule_id': 'MERRA2_400.tavg1_2d_slv_Nx.20210601.nc4',
+            'boxes': ['-90 -180 90 180'],
+            'time_start': '2021-06-01T00:00:00.000Z'}]
+        
+        granule_response_no_producer_id = [{
+            'boxes': ['-90 -180 90 180'],
+            'time_start': '2021-06-01T00:00:00.000Z'}]
+        
+        with self.subTest('Granule has "producer_granule_id"'):
+            self.assertEqual(get_granule_name(granule_response),
+                             'MERRA2_400.tavg1_2d_slv_Nx.20210601.nc4')
+        
+        with self.subTest('Granule does not have "producer_granule_id"'):
+            with self.assertRaises(IndexError):
+                get_granule_name(granule_response_no_producer_id)
+    
+    
+    def test_download_gran_exceptions(self):
+        ''' Check if exceptions are raised for incorrect input parameters
+        '''
+            
+        with self.assertRaises(ValueError):
+            # Input parameter, `granule_link` should be string
+            download_granule(5, 'foo')
+            
+        with self.assertRaises(ValueError):
+            # Input parameter, `out_filename` should be string
+            download_granule('foo', 10)
+        
+        with self.assertRaises(ValueError):
+            # Input parameter, `out_filename` should be string
+            download_granule('foo', 10)
+        
+        with self.assertRaises(RequestsException):
+            # Not a real https link
+            download_granule('foo', 'foo.nc4')
+    
+    
+    def _mock_requests(self, status=200, content="CONTENT"):
+        '''
+        Mock requests.get module and it's methods content and status.
+        Return the mock object.
+        '''
+        mock_response = mock.Mock()
```

Comment:
Sounds good, do you think you could elaborate on why I would do this? I just want to make sure I am following.

Reply:
I think this is mainly a distinction between the function being mocked and the object being returned. You are mocking `requests.get`, which makes a lot of sense - we don't want to attempt to make HTTP requests during the tests, and we want to be able to establish `requests.get` was called with the correct parameters.

Generally, it's nice to make the return value of the `Mock` as close to the real thing you'd expect to get as possible, in this case a `requests.Response` object. You can do that by creating a new instance of an object and setting its attributes to what your code expects, so that when you test the code it can process that object as though things are normal.

The tricky thing is that it turns out creating a `requests.Response` instance isn't as easy as it could be. I'm sure there's a way, but I've not found it from a quick bit of playing around. Given that it is a bit trickier, the short cut here is to use `mock_response = Mock(spec=requests.Response)`, so at least the output object from this function has all the right attributes and methods, even if it isn't really a `requests.Response`.

(Sorry for the essay!)